src/System.IO.FileSystem/src/System/IO/UnixFileSystem.cs

```diff
@@ -504,6 +505,17 @@ private sealed class FileSystemEnumerable<T> : IEnumerable<T>
                         }
                         searchPattern = searchPattern.Substring(lastSlash + 1);
                     }
+
+                    // Typically we shouldn't see either of these cases
+                    if (searchPattern.IndexOfAny(s_EscapeCharacters) != -1)
```

Comment:
It allocates an array for indexes in the native code I believe. I didn't look super closely at it- I might have misread it. In any case it is a fair amount of logic.

As far as a general improvement I don't know what we can do- the advantage here is that we have context- we normally expect there to be no match.